package exporter ui simplification + metainfo for packages in config.lua

Edru2 opened the pull request:
#### Brief overview of PR changes/additions
Changes the ui for the package exporter to simplify package creation and allowing to add (optional)
metainformation to a package which will be saved to config.lua and can be get by getter functions.
Xml packages which don't come with a config.lua can save the metainfo by using setter functions.
The metainformation are also saved as directly in the mpackage zip file as  zip file comment which could
make it easier in future to read them directly and autofill the settings in the package repo.

MetaInformations will be saved in mudlet.packageInfo/mudlet.moduleInfo respectively.
new getter setter functions are:
*getPackageInfo(packageName, [info])*
*getModuleInfo(moduleName, [info])* if info is not given the whole meta-informations for that package will be returned as table
*setPackageInfo(packageName, info, value)*
*setModuleInfo(moduleName, info, value)*  if package/moduleName has no entry in mudlet.package/moduleInfo the entry will be created
example
```lua
setPackageInfo("myPackage", "author", "Edru")
author = getPackageInfo("myPackage","author")
echo(f("\nThe author of the package 'myPackage' is {author}."))
-- this will return: The author of the package 'myPackage' is Edru.
```

The new package-exporter ui additionally allows.
-Creating new packages based on installed packages.
-After exporting a package the package exporter doesn't need to be restarted to create another package
-Files can be added by drag and drop and/or by selecting folders/files in a dialog

#### Motivation for adding to Mudlet
I imagine this can help a lot with version handling in the package repo.
It also should allow for easier package exporting.
fix #4870
fix #2675

#### Other info (issues closed, discussion etc)
I already work on changes for the package manager to visualize the meta informations at https://github.com/Edru2/Mudlet/tree/new-package-manager-test

#### Release post highlight
-Package exporter ui overhaul to simplify package creation and adding also the possibility for meta-informations (Author, Icon, Description, Version...) included in a package

Edru2 committed:
get and fill fillPackage info

Edru2 committed:
global Infos instead of _G

Edru2 committed:
use lua to do the file handling

Edru2 committed:
first try on packageinfo input

it is not functional yet just something to look at ;)

SlySven commented on src/Host.cpp:
```diff
@@ -1913,73 +1913,77 @@ bool Host::uninstallPackage(const QString& packageName, int module)
+
+QString Host::getPackageConfig(const QString& luaConfig, int isModule)
+{
+    QString packageName{QString()};
```
This is pointless, the default constructor for a QString gives an empty one, just use:
```suggestion
    QString packageName;
```

SlySven commented on src/Host.cpp:
```diff
@@ -1913,73 +1913,77 @@ bool Host::uninstallPackage(const QString& packageName, int module)
     luaL_openlibs(L);
-
-    int error = luaL_loadstring(L, strings.join("\n").toUtf8().constData());
+    int error = luaL_loadfile(L, luaConfig.toUtf8().constData());
```
This could break on Windows as it won't work if there are accented characters in the path or file name - which is why we read the file contains with `QFile` which can worked with whatever local8Bit encoding is used for file names - the `lua_loadfile(...)` uses `std::iostream` IIRC which doesn't...

Edru2 replied:
Thanks for the information I wasn't aware that this works different on Windows. 
I just tested it and it indeed doesn't work in Windows when using the package name ```DárkThéöä黑.mpackage``` 
I also noticed that using the same package name in Mudlet 4.10.1 works but changes the package name to 
![image](https://user-images.githubusercontent.com/60551052/109383149-de286d00-78e4-11eb-826f-622e5109dc39.png)
but still better then not working at all ;) I will revert it

SlySven commented on src/Host.cpp:
```diff
@@ -1913,73 +1913,77 @@ bool Host::uninstallPackage(const QString& packageName, int module)
+    }
+
+
+    if (mudlet::debugMode) {
```
This isn't an appropriate test IMO - it says whether the Central Debug Console is active whereas the error message is sent to standard ouput (i.e. the command line).

SlySven commented on src/TLuaInterpreter.cpp:
```diff
@@ -13898,6 +13939,31 @@ void TLuaInterpreter::setupLanguageData()
+    if (!lua_istable(L, 1)) {
+        return;
+    }
+    lua_newtable(L);
```
What happens if there is more than one package or module? I am not sure but don't you only have to create the `mudlet.packageInfo` and `mudlet.moduleInfo` tables in the main Lua state if they do not already exist - I can't see if that is what you are doing or not. In Lua it would be:
```lua
mudlet = mudlet or {}
mudlet.packageInfo = mudlet.packageInfo or {}
mudlet.moduleInfo = mudlet.moduleInfo or {}
```
that being the case it might be easer just to do this where the `mudlet = mudlet or {}` is done elsewhere

Edru2 replied:
Maybe I should clear about what this does in a comment? loadPackageInfos happens only once when the profile gets loaded and then loads all the config.lua files which can be found in the profile folder. That means it only works for packages as modules work different and get installed every time Mudlet reloads.
The mudlet table is already created in setupLanguageData() and the 
```cpp    
 if (!lua_istable(L, 1)) {
        return;
    }
```
part is probably not needed (will get rid of it)
moduleInfo is just created but not really used here.

This is already tested and works with one or multiple packages.

>  do this where the mudlet = mudlet or {} is done elsewhere

That  happens in  setupLanguageData() and it feels wrong to create packageInfo and moduleInfo there as this has nothing to do with language data.

and as Vadi already mentioned this is still a WIP and unfinished there are many issues I'm aware of and still have to solve ;)

SlySven replied:
Fair enough - but for some of the other issues it probably helps if other developers point out things that might bite you...!

SlySven commented on src/TLuaInterpreter.cpp:
```diff
@@ -13898,6 +13939,31 @@ void TLuaInterpreter::setupLanguageData()
+    QStringList packages = mpHost->mInstalledPackages;
+    for (int i = 0; i < packages.size(); i++) {
+        QString packagePath{mudlet::self()->getMudletPath(mudlet::profilePackagePath, mpHost->getName(), packages.at(i))};
+        QDir _dir(packagePath);
```
Don't use leading `_` on identifiers - whilst **single** ones are okay PROVIDED THEY ARE NOT FOLLOWED BY CAPITAL CASE letters it is not something that is wise and we should be moving away from it...

SlySven commented on src/dlgPackageExporter.cpp:
```diff
@@ -170,6 +181,39 @@ bool dlgPackageExporter::writeFileToZip(const QString& archiveFileName, const QS
+{
+    if(evt->type() == QEvent::KeyPress) {
+        QKeyEvent *keyEvent = static_cast<QKeyEvent*>(evt);
+        if(keyEvent->key() == Qt::Key_Escape || keyEvent->key() == Qt::Key_Return || keyEvent->key() == Qt::Key_Enter)
```
Code styling matter: we use `{`...`}` even on single lijne `if` code blocks.

SlySven commented on src/ui/dlgPackageExporterInput.ui:
```diff
@@ -0,0 +1,230 @@
+   <item>
+    <widget class="QWidget" name="widget" native="true">
+     <layout class="QGridLayout" name="gridLayout">
+      <item row="3" column="0">
```
If you can edit (in an editor with code folding makes it easier, like the *plain text editor* in Qt Creator) these entries to be in ascending row / column order then it reduces the noise in Git later on if changes are made - this is because the Qt Designer plugin/utility does NOT keep `QGridLayout` items sorted (or bother to remove completely empty columns or rows, e.g. you only have columns `0` and `5` now - and none of them span more than one column so you can clean it up by changing all the `column="5"` entries to be `column="1"` ones.

SlySven replied:
:information_source: Actually I looked at this locally and found that the two column layout with labels on the left and entry widgets on the right is ideally handled by the `QFormLayout` and it is simple in Qt Designer to change the layout to that type in a single step...! :grinning:

SlySven commented on src/dlgPackageExporter.cpp:
```diff
@@ -170,6 +181,39 @@ bool dlgPackageExporter::writeFileToZip(const QString& archiveFileName, const QS
+    input->Dependencies->clearEditText();
+}
+
+bool dlgPackageExporter::eventFilter(QObject* obj, QEvent* evt)
```
I think you need to test the `obj` value to be certain it applies to the right *object* - possibly that would be the `input` one?
:man_shrugging:

vadi2 commented:
I believe the PR marked as draft and WIP

Edru2 committed:
repair removePackageInfo + add suggestions

add SlySven suggestions

Edru2 committed:
add possibility to add infos through gui

also informations are add to zip comment
and package exporter gui is fully usable by keyboard

Edru2 committed:
copy paste fail

dependencies were called author

Edru2 committed:
Update dlgPackageExporter.cpp

Edru2 committed:
add include

SlySven commented on src/Host.cpp:
```diff
@@ -1824,7 +1824,11 @@ bool Host::uninstallPackage(const QString& packageName, int module)
         }
     }
-
+    //module == 2 seems to be only used for reloading/syncing, which doesn't work for mpackages anyway
```
Yeah - for reference the values are - I *believe*:

0. Load as a package
1. Load as a module (from the module/package managers)
2. (Re)load as a module being synced (updated in another profile and then reloaded to get the new one)
3. Load as a module (from a Lua script)

Edru2 committed:
make dependecies more intuitive

Edru2 committed:
error handling and some comments

Edru2 committed:
Update Host.cpp

Edru2 committed:
allow multiline descriptions

Edru2 committed:
support markdown language in description

Not sure if we should make it optional

vadi2 commented on src/dlgPackageExporter.cpp:
```diff
@@ -281,12 +287,27 @@ bool dlgPackageExporter::eventFilter(QObject* obj, QEvent* evt)
+
+        if (evt->type() == QEvent::FocusOut) {
+            mPlainDescription = input->Description->toPlainText();
+            input->Description->setMarkdown(mPlainDescription);
```
It's OK that Qt 5.11 doesn't support markdown - all official releases are on a more modern Qt that does. For the older Qt you can display markdown as-is, it's a compréhensible format.

Edru2 committed:
add $profilePath parameter in description

this can facilitate  images in description

Edru2 committed:
Merge branch 'development' into get-info-from-package

Edru2 committed:
avoid duplicates in dependencies

This avoids duplicates in dependencies + reads the dependency textedit and adds the element if it is not added on 'OK'.
Also cleans up the inputDialog eventfilter (which is not needed anymore)

Edru2 commented on src/dlgPackageExporter.cpp:
```diff
@@ -170,6 +209,128 @@ bool dlgPackageExporter::writeFileToZip(const QString& archiveFileName, const QS
+bool dlgPackageExporter::eventFilter(QObject* obj, QEvent* evt)
+{
+    if (evt->type() == QEvent::KeyPress) {
+        QKeyEvent* keyEvent = static_cast<QKeyEvent*>(evt);
```
I read on https://forum.qt.io/topic/107776/static_cast-for-downcasting-event/6 that static_cast is cheaper the dynamic_cast and dynamic_cast is not really needed in this case as it is checked beforhand if the event is a qkeyevent.
The same for the warning below.

vadi2 replied:
Yep good catch. This check will be disabled with https://github.com/Mudlet/Mudlet/pull/4903/files

Edru2 committed:
add Title and Icon

This change allows to add an Icon and short description (title)
which then can appear directly in the packagemanager

vadi2 committed:
Merge remote-tracking branch 'origin/development' into get-info-from-package

Edru2 committed:
new package_exporter ui

vadi2 committed:
Merge remote-tracking branch 'origin/development' into get-info-from-package

Edru2 committed:
folder copy (drag and drop)

folder copy working now with drag and drop
adds ability to select modules in the package export list
when a package is selected autofill values

Edru2 committed:
codefactor

Edru2 committed:
add QMimeData so it compiles on older qt version

Edru2 committed:
codefactor another go

Edru2 committed:
allow selections of (multiple) folder and files in addFiles dialog

Edru2 committed:
simplify version add get and set package and module info

Edru2 committed:
export multiple packages

Edru2 committed:
title was written wrong everywhere

vadi2 commented on src/dlgPackageExporter.cpp:
```diff
@@ -511,6 +841,7 @@ void dlgPackageExporter::slot_export_package()
                 // details):
                 // Change the cursor to a system busy one while we are working:
                 QApplication::setOverrideCursor(Qt::BusyCursor);
+                zip_set_archive_comment(archive, mPackageConfig.toUtf8().constData(), mPackageConfig.length());
```
This is an excellent idea!

Just leaving a note here - in case the zip archive doesn't have a comment (if it wasn't made by Mudlet), we'll have to still extract it to read the config.lua inside

Edru2 replied:
My idea when adding this was to allow the package repo in future to directly read the package info (and then autofill it in the forms) without needing to extract the whole package.
Mudlet still only reads the info from the config.lua

Kebap commented:
> title was written wrong everywhere

Haha, Title is German for title 😉 That's our German development heritage

SlySven commented on src/ui/dlgPackageExporter.ui:
```diff
@@ -90,36 +573,51 @@
  <connections>
   <connection>
    <sender>buttonBox</sender>
-   <signal>accepted()</signal>
```
The changes on and below this line are just Git noise from the Qt designer just randomly swapping two signal-slot connections that are made from this file (rather than in the CPP class that has the slots) - I am not sure that all us developers are aware that such things which are equivalent to a `QObject::connect(...)` call in our C++ code can happen - but they can! :smiling_imp:

As it happens such connections also show up in one of the (lesser used) view tabs in the Qt Designer tool/plugin - IIRC this may or may not be the same one that shows buddy connections (click on a label for the focus to jump to the buddy) between labels and the field that they are buddy's with...

vadi2 commented on src/TLuaInterpreter.cpp:
```diff
@@ -13070,6 +13070,75 @@ int TLuaInterpreter::unzipAsync(lua_State *L)
 }
 
+// No documentation available in wiki - internal function
+void TLuaInterpreter::fillPackageInfo(const QString& packageName, bool isModule, lua_State* packageLua)
```
What do you think about just returning this info on-demand as a Lua function? Maintaining this data in the Lua state wouldn't be consistent with the rest of the API, and someone is bound to try to change it in Lua - expecting it to take effect.

Edru2 replied:
>  change it in Lua - expecting it to take effect.

If someone changes it in Lua it will take effect see #4938.
The functions setPackageInfo/setModuleInfo will also allow to add custom infos for packages/modules which are pure xml and therefore don't come with a config.lua.
For custom infos (for example Homepage or License) check the example package at #4938 

I'm obviously open for better ideas as I'm also not 100% happy with this.

vadi2 committed:
Improvements to package exporter (#9) by vadi2

* Add count to items

* Count just once for multiple selections

* Add placeholder default location for convenience

* Update placholder

* Tweak .ui

* WIP

* Simplify UI

* Enable/disable button as necessary

* Incorporate Leris' feedback

vadi2 commented:
Something I'd like to have still is a dashed border for the area where we can drag and drop files:
![image](https://user-images.githubusercontent.com/110988/111764147-2fec6380-88a3-11eb-93fe-5c97ae03ad26.png)

Otherwise it's all good on my side. @Kebap has some feedback he'd like to post as well, and then we are done. This PR has been outstanding for a while, should be plenty of time for review.

vadi2 commented:
Remembered another thing: the temporary staging directory isn't cleared/reset, so when you restart Mudlet and try to re-export the same package, the UI won't show that it'll really export in the end.

Edru2 committed:
don't use lua api + change for dependecies + use os tempDir

Edru2 committed:
codefactor + clang-tidy

vadi2 committed:
Merge remote-tracking branch 'origin/development' into get-info-from-package

vadi2 committed:
Add border to drag and drop area

vadi2 commented:
Added borders for drag and drop:

![image](https://user-images.githubusercontent.com/110988/111861607-f160b280-894f-11eb-81fa-5773d1e20517.png)
![image](https://user-images.githubusercontent.com/110988/111861609-f45ba300-894f-11eb-9068-7f90b86180f4.png)

vadi2 committed:
Rename dependencies to required packages

Non-software folk don't know what a dependency is in this context

vadi2 committed:
Don't apply dashed border to tooltip

vadi2 commented:
Ideas for improvement in future PRs:

![image](https://user-images.githubusercontent.com/110988/111862065-130f6900-8953-11eb-9abf-721d9ec06685.png)

Edru2 committed:
index should be 0 at for loop

vadi2 committed:
Change Icon to .mudlet/Icon

vadi2 committed:
Use proper ISO8601 timestamp

This'll make it simple because it's the local time, and it'll work in our international package manager as well because the timezones are accounted for

vadi2 commented:
Tickets:

https://github.com/Mudlet/Mudlet/issues/4993
https://github.com/Mudlet/Mudlet/issues/4994
https://github.com/Mudlet/Mudlet/issues/4995
https://github.com/Mudlet/Mudlet/issues/4996
https://github.com/Mudlet/Mudlet/issues/4997

vadi2 committed:
Rename UI elements to follow Qt convensions

SlySven introduced them and I have to say they do look nicer and cleaner

vadi2 reviewed (approved):
Looks good to me. Have ideas for improvement, but I'm confident we can get them in before release and this is good to be seen by the PTB testers now.

vadi2 commented:
Need to document that we're using github-flavoured markdown.
